File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def removeNthFromEnd(self, head, n):
        """
        :type head: Optional[ListNode]
        :type n: int
        :rtype: Optional[ListNode]
        """
        if not head:
            return None
        
        cnt, temp = 0, head
        while temp:
            cnt += 1
            temp = temp.next
        
        if cnt == n: return head.next

        res = cnt - n
        temp = head

        while res > 1:
            res -= 1
            temp = temp.next
        
        temp.next = temp.next.next
        return head

File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        if(head == null){
            return null;
        }

        int count = 0;
        ListNode temp = head;

        while(temp != null){
            count += 1;
            temp = temp.next;
        }

        if(count == n){
            return head.next;
        }

        int res = count - n;
        temp = head;

        while(res > 1){
            res -= 1;
            temp = temp.next;
        }

        temp.next = temp.next.next;
        return head;
    }
}